### PR TITLE
Add jsonEscape option to generator

### PR DESCRIPTION
**Pull request by kangax**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | yes
| Deprecations?     | no
| Spec compliancy?  | yes
| Tests added/pass? | yes
| License           | MIT

This adds an option to configure the way non-ASCII safe characters are encoded in Babel. "\x.." notation is shorter but it not compatible with some of the parsers.

### CODE REVIEW

File: packages/babel-generator/src/index.js [hunk 1]
```diff
@@ -58,6 +58,7 @@ function normalizeOptions(code, opts, tokens): Format {
     minified: opts.minified,
     concise: opts.concise,
     quotes: opts.quotes || findCommonStringDelimiter(code, tokens),
+    jsonEscape: opts.jsonEscape,
```
Comment (kangax) [inline packages/babel-generator/src/index.js]: Not very happy with this name. Any other options?

Comment (mathiasbynens) [inline packages/babel-generator/src/index.js]: `jsonCompatibleStrings`?

**Comment by loganfsmyth** [comment]
It seems like maybe an `encoding: 'ascii'` option would be more readable. Thoughts?

**Comment by kangax** [comment]
I like "encoding: ascii" but what would the other value be?

**Comment by mathiasbynens** [comment]
> It seems like maybe an `encoding: 'ascii'` option would be more readable. Thoughts?

I don’t understand. Babel’s output for string literals is ASCII-safe either way, with or without this new `jsonEscape` option enabled. It’s just that the escape sequences are either as compact as possible (`jsonEscape: false`) or JSON-compatible (`jsonEscape: true`). Could you clarify what you mean?

**Comment by kangax** [comment]
@mathiasbynens you're right, `jsonCompatibleStringLiterals` makes more sense then

**Comment by kangax** [comment]
@hzoo suggests `jsonCompatibleStrings` which might be even better (don't see a good reason to be so specific)

**Comment by loganfsmyth** [comment]
I don't follow this `json` argument name, can we reconsider? having non-ASCII characters in string literals is allowed by JSON. If the expectation is to have `ascii` or `utf8` output, can we make ane `encoding` argument that defaults to `utf8`?

**Comment by loganfsmyth** [comment]
Sorry, ignore my previous comments, I was just being dumb and misunderstanding what this was aiming to achieve.

Having an argument for this seems fine, but could we clarify somewhere what the actual usecase of this argument is? @kangax you mentioned `"\x.." notation is shorter but it not compatible with some of the parsers.`, what doesn't support `\x`?

**Comment by mathiasbynens** [comment]
@loganfsmyth Some context here: https://github.com/babel/babel/pull/4478#r82436823 It seems Facebook is using PHP’s `json_decode()` with a string literal output by Babel as its argument.

**Comment by loganfsmyth** [comment]
@mathiasbynens Perfect, I'll shut up now :P

**Pull request merged**